extensions/mapstore/src/main/java/com/hazelcast/mapstore/GenericMapStore.java

```diff
@@ -0,0 +1,518 @@
+                logger.fine("Discovered following mapping columns: " + mappingColumns);
+            }
+
+            try (SqlResult ignored = sql.execute(
+                    "CREATE MAPPING \"" + mapping + "\" "
+                            + "EXTERNAL NAME \"" + properties.tableName + "\" "
+                            + (mappingColumns != null ? " ( " + mappingColumns + " ) " : "")
+                            + "TYPE " + deriveMappingType() + " "
+                            + "OPTIONS ("
+                            + "    '" + OPTION_EXTERNAL_DATASTORE_REF + "' = '" + properties.externalDataStoreRef + "' "
+                            + ")")) {
+            }
+
+            if (!properties.hasColumns()) {
+                columnMetadata = loadMetadataFromMapping(mapping).getColumns();
+            }
+            queries = new Queries(mapping, properties.idColumn, columnMetadata);
+        } catch (Exception e) {
+            if (e.getMessage() != null && e.getMessage().startsWith("Mapping or view already exists:")) {
+                readExistingMapping();
+                return;
+            } else {
+                logger.warning(e);
+                initFailure = e;
+            }
+        }
+
+        initFinished.countDown();
+    }
+
+    private String deriveMappingType() {
+        if (properties.mappingType != null) {
+            return properties.mappingType;
+        } else {
+            ExternalDataStoreFactory<?> factory = nodeEngine()
+                    .getExternalDataStoreService()
+                    .getExternalDataStoreFactory(properties.externalDataStoreRef);
+
+            if (factory instanceof JdbcDataStoreFactory) {
+                return "JDBC";
+            } else {
+                throw new HazelcastException("Unknown ExternalDataStoreFactory class " + factory.getClass()
+                        + ". Set the mapping type using '" + MAPPING_TYPE_PROPERTY + "' property");
+            }
+        }
+    }
+
+    private String resolveMappingColumns(String tableName, String externalDataStoreRef) {
+        String tempMapping = "temp_mapping_" + UuidUtil.newUnsecureUuidString();
+        createMapping(tempMapping, tableName, externalDataStoreRef);
+        SqlRowMetadata rowMetadata = loadMetadataFromMapping(tempMapping);
+        columnMetadata = rowMetadata.getColumns();
+        dropMapping(tempMapping);
+
+        return Stream.concat(of(properties.idColumn), properties.columns.stream())
+                     .distinct() // avoid duplicate id column if present in columns property
+                     .map((columnName) -> {
+                         int column = rowMetadata.findColumn(columnName);
+                         if (column == COLUMN_NOT_FOUND) {
+                             throw new HazelcastException("Column '" + columnName + "' not found");
+                         }
+                         return column;
```

Comment:
Maybe extract `validateColumn` and reuse it in `validateColumns`

Reply:
5f4cd762